diff --git a/index.js b/index.js
index 0ed8554d4ac1212086a0968eb80b172bea65b5a2..9ced44a71e1ab5154ea39e63028d180396a9b088 100644
--- a/index.js
+++ b/index.js
@@ -3,10 +3,10 @@
  * @author imcuttle
  */
 
-const visit = require('unist-util-visit')
-const { setNodeId, getDefaultId } = require('./util')
+import { visit } from 'unist-util-visit'
+import { setNodeId, getDefaultId } from './util'
 
-module.exports = function(options = { defaults: false }) {
+export default function(options = { defaults: false }) {
   return function(node) {
     visit(node, 'heading', node => {
       let lastChild = node.children[node.children.length - 1]
diff --git a/package.json b/package.json
index 3a0e06961eb18ce998d7ffef3675584e8f4ff28c..8b3c287e9d2879b48034be348fdf467347c6d7b2 100644
--- a/package.json
+++ b/package.json
@@ -1,5 +1,6 @@
 {
   "name": "remark-heading-id",
+  "type": "module",
   "version": "1.0.1",
   "main": "index.js",
   "description": "The remark plugin for supporting custom heading id also including default id",
diff --git a/util.js b/util.js
index 6255a3146d666077bad3e9b077814c24b2b63d52..92d90a88d15181f1d231c4565988eca6f441d9f5 100644
--- a/util.js
+++ b/util.js
@@ -1,9 +1,9 @@
-const _ = require('lodash')
+import { isEmpty, kebabCase } from 'lodash'
 
-const extractText = children => {
+export const extractText = children => {
   return children
     .map(child => {
-      if (!_.isEmpty(child.value)) {
+      if (!isEmpty(child.value)) {
         return child.value
       } else if (child.children && child.children.length > 0) {
         return extractText(child.children)
@@ -14,18 +14,16 @@ const extractText = children => {
     .join(' ')
 }
 
-const getDefaultId = children => {
+export const getDefaultId = children => {
   return formatDefaultId(extractText(children))
 }
 
-const formatDefaultId = value => {
-  return _.kebabCase(value.replaceAll(/\\s+/g, ' ').trim())
+export const formatDefaultId = value => {
+  return kebabCase(value.replaceAll(/\\s+/g, ' ').trim())
 }
 
-const setNodeId = (node, id) => {
+export const setNodeId = (node, id) => {
   if (!node.data) node.data = {}
   if (!node.data.hProperties) node.data.hProperties = {}
   node.data.id = node.data.hProperties.id = id
 }
-
-module.exports = { extractText, getDefaultId, formatDefaultId, setNodeId }
